### DISCUSSION TITLE
Issue Again with no Scrollbar, and with Menu Cutting Off

### DISCUSSION THREAD

**Discussion by NCIDevRonR** [primefaces/community#33] [PrimeFaces] [answered]
I'm heading up our project team's migration from PrimeFaces 6 to PrimeFaces 13, and it looks like the normal vertical scrollbar which appears when the user zooms in too closely, no longer appears.  I've looked into scroll panels and overflow properties in divs, but they don't behave the same as a whole screen scroll bar.

Our content page didn't have or need any special code to enable whole-page scrolling under PrimeFaces 6.  Is there a way to accomplish this in PrimeFaces 13?

**Comment by FlipWarthog** [comment]
Unfortunately, a lot has changed between PrimeFaces 6 and PrimeFaces 13, so it's hard to track down what specifically is causing the behavior difference you're describing. I would suggest setting up a reproducer for comparison using the [primefaces-test](https://github.com/primefaces/primefaces-test) project which shows what you're describing in both PF6 and PF13. This will greatly help in being able to track down the specific cause/solution. Note, if you do set up a reproducer for PF6, you'll need to change the theme to one that existing in PF6, as well as change the `primefaces` dependency version, e.g.:
![image](https://github.com/primefaces/community/assets/83613837/93491daf-479d-4912-b5b4-263d798f4fe2)

**Reply by NCIDevRonR** [reply]
Thanks for your reply.  The test project sounds like a good idea, but I'm not having the best luck it.  I've cloned it, and built it in NetBeans 13, but I get an IllegalStateException when I try to deploy it to my JBoss 7.1 instance under Java jdk8_371 and Windows 10.  I'm attaching the stack trace.
[PrimeFaces Test Attempt-20230914.txt](https://github.com/primefaces/community/files/12610212/PrimeFaces.Test.Attempt-20230914.txt)

**Reply by FlipWarthog** [reply]
Can you try running it with Jetty from the command line with `mvn clean jetty:run`?

**Reply by NCIDevRonR** [reply]
I've got the little welcome page.  How do I fill up a screen full of stuff to test the scroll bars?

![Jetty-PF-Test-Basic-20230914-001](https://github.com/primefaces/community/assets/78611742/ef4c28b2-f83b-4b5b-905f-282bfd1d2dea)

**Reply by FlipWarthog** [reply]
Awesome, you can now add whatever you need to reproduce the issue into test.xhtml and TestView.java.

**Reply by NCIDevRonR** [reply]
In the test project, it appears to be working...
![image](https://github.com/primefaces/community/assets/78611742/f3f395f2-355a-4bca-ae3d-6ee9dd5e498f)
...which means my project isn't set up the same way.

![Jetty-PF-Test-Dependencies-20230914-001](https://github.com/primefaces/community/assets/78611742/7a6f5e48-c88b-485c-934d-f1ee6c9105dc)
I'm going to look into incorporating some (or all) of the dependencies here that I don't have in my project.  Any idea which ones handle scrollbars?

**Reply by jepsar** [reply]
Note that you are probably using a different [theme](https://primefaces.github.io/primefaces/13_0_0/#/core/themes).

**Comment by cagataycivici** [comment]
Do you have a visual sample like a screenshot comparing two?

**Reply by NCIDevRonR** [reply]
Here are some visual examples, where I've zoomed in the screen.  As you can see, the new screen doesn't have a vertical scroll bar, whereas the old screen does.
![Old Screen Zoomed In-20230914](https://github.com/primefaces/community/assets/78611742/a7b5c123-4d76-456f-b5e3-ec022a88991e)
![New Screen Zoomed In-20230914](https://github.com/primefaces/community/assets/78611742/1eaba070-6db9-404d-942a-9c412852dcd4)

**Reply by melloware** [reply]
Are you using an old JQuery ThemeRoller theme?  that will most likely be your issue?

**Reply by NCIDevRonR** [reply]
I have been using "smoothness". I have tried switching to "saga" and "vela", and still the issue persists. Interestingly though, in "vela" theme, which is decidedly darker than the others, when I open my zoomed-in screen, I see a vertical scrollbar flash on, then disappear, so it's apparently there, but is somehow disabled. Does any of this make sense?
![Vela Screen Zoomed-20230915-001](https://github.com/primefaces/community/assets/78611742/93a67217-b67b-4ad8-b17b-5e9990ce458a)

**Reply by melloware** [reply]
Check any custom CSS you may be using also in your app.

**Reply by NCIDevRonR** [reply]
I'm attaching our custom css we've been using since day 1.  To me it looks fairly straight-forward, as a non-css expert.  I've recently also begun using primeflex.css, to try and get back some control over field widths and columnar formatting.

Could something in primeflex.css be the issue?
[my_custom_css.txt](https://github.com/primefaces/community/files/12623345/my_custom_css.txt)

**Reply by melloware** [reply]
I see these two dealing with Overflow? Are you using pe:layout ?

```js
.ui-layout-north {
    z-index: 20 !important;
    overflow: visible !important;;
}
.ui-layout-north .ui-layout-unit-content {
    overflow: visible !important;
}
```

Also all I can do is speculate if you can create a PrimeFaces Test reproducer showing the problem I could see it in my browser and debug it.

**Reply by NCIDevRonR** [reply]
I created a PF test app and loaded it to the gills with elements, and the scrollbars appeared and worked just fine.

We are using pe:layout in our screen template file.  I'm attaching it here.
[mytemplate_xhtml.txt](https://github.com/primefaces/community/files/12623509/mytemplate_xhtml.txt)

**Reply by melloware** [reply]
OK if I had to guess its `pe:layout` related since that is controlling that "pane"/"scrollbar" and you might want to update PrimeFaces Test sample to use Layout so its apples to apples.

**Reply by melloware** [reply]
PFE showcase uses Layout: https://www.primefaces.org/showcase-ext/views/home.jsf

And I see `overflow:auto`

![image](https://github.com/primefaces/community/assets/4399574/90cfc898-bdee-45f0-9edc-582599273040)

IF I edit that and remove the overflow:auto in Chrome Tools I see my scrollbar dissappears.

**Reply by FlipWarthog** [reply]
@melloware I think you're right, looks like it's falling back to this:
```css
.ui-layout-pane {
    padding: 0;
    overflow: hidden;
}
```

The PFE showcase has some extra styles in it, but updating the primefaces-test example will give a clearer view of what's going on and what needs to change to correct. Should just be a slight modification of your custom CSS.

**Reply by NCIDevRonR** [reply]
I've added our CSS files to the PF test's test.xhtml's list of CSS files, and I've changed the overflow setting to "hidden", but I'm so far, unable to see the PF test app break.  The vertical scrollbar has worked every time I've tried it today.

**Reply by melloware** [reply]
@NCIDevRonR did you add PE Layout to you Test project?

**Reply by melloware** [reply]
Any update @NCIDevRonR ?

**Reply by NCIDevRonR** [reply]
Here's the link to my vertical scrollbar issue reproducer, which as yet still doesn't capture the issue:  [https://github.com/NCIDevRonR/primefaces-test/tree/no-vertical-scrollbar](https://github.com/NCIDevRonR/primefaces-test/tree/no-vertical-scrollbar)  I'll keep trying to recreate it in here, and am meanwhile open to suggestions for possible causes and fixes.

**Reply by NCIDevRonR** [reply]
I have just added the PrimeFaces Extensions to the pom.xml, and am now getting the scrollbar issue in this branch of my repo, along with this message in the GUI.
![Center Pane Element Error](https://github.com/primefaces/community/assets/78611742/789bb731-ecda-46c1-ba56-599f00ed6707)
I'm not sure if this is an exact duplicate of what's going on in our app, but at least it's very similar.  Any ideas folks?

**Reply by melloware** [reply]
Yep fixed your example: 
[primefaces-test-no-vertical-scrollbar.zip](https://github.com/primefaces/community/files/12719361/primefaces-test-no-vertical-scrollbar.zip)

The error was correct a Layout required a Center Pane.  Also you were missing the Google GSON jar in your pom.xml.  Once I fixed both your example is working.

**Reply by NCIDevRonR** [reply]
Thanks for this fix.  I've committed it to my repo.  Trouble is, that version of the reproducer didn't look much like our application.

...so I'm bringing in a little more of what our application actually looks like code-wise, because I have about 30 screens formatted the same way which will need whatever updating ends up working for this one.  In my reproducer I seem to be having a more basic problem.  For some reason, my example screen shows up blank.  What do I need to make this screen show up?

I'm also including photos of what our users are currently seeing in PF6, vs what our application looks like in PF13.

In the PF13 photo, you'll notice there is no scroll bar.  Will the fix you arrived at for my simplistic screen work for our actual screen?  

What my team would like is for this screen under PF13 to look as much as possible like it currently does under PF6, with the fields and labels close together in nice orderly grids and sections for Person, User and User-to-Group.  I've chased around from doc page to doc page looking for something which will address this.  Any suggestions?

![MyApp - Old School](https://github.com/primefaces/community/assets/78611742/506ba65c-2d47-4921-b729-5e8558443d32)
![MyApp - Flex Divs](https://github.com/primefaces/community/assets/78611742/5fc869a6-34f8-434d-8903-966f96004ce8)

...so basically I'm looking to resolve a blank screen, some formatting, and the scrollbar, and still keep as much of the existing layout as possible, so I can replicate the change over our template and our 30 or so screens.  Any suggestions?

**Reply by melloware** [reply]
Lots of issue here. First if you want your panelGrid to looks like it did do 2 things 1 add this CSS and 2 add this to your panel grids `layout="tabular" columns="0"`

```css
.ui-panelgrid .p-grid .ui-panelgrid-cell, .ui-panelgrid .grid .ui-panelgrid-cell, .ui-panelgrid .ui-grid-responsive .ui-panelgrid-cell {
    border: 1px solid #dee2e6;
}
```

That first snippet is from the migration guide: https://primefaces.github.io/primefaces/13_0_0/#/../migrationguide/13_0_0?id=panelgrid

**Reply by NCIDevRonR** [reply]
In order to get this panelGrid fix to work, I ended up having to specify columns="1", and to add p:row elements to delineate which fields belong on individual rows.  Here's how it looks now.
![MyApp - FormatFix - 20230927-001](https://github.com/primefaces/community/assets/78611742/306e69ea-927e-44ea-bd55-fe45c16f375d)
I got crazy results when I tried to add p:column elements and change to different values of columns=?, so for now, I'm going to run this by my guys as the closest I can get to how it currently is in prod.
Next, I'm going to try incorporating the pe:layout/pe:layoutPane fix you gave me yesterday into our application to get back our scrollbars.  I'm guessing I'll need to scale this out to all of our screens which are prone to overflowing.  I'll let you know how it goes with this one screen.

**Reply by NCIDevRonR** [reply]
...so here's how it looks with the addition of the pe:layoutPane elements.  I had to replace the **ui:composition** and **ui:define** elements with **html**, **head**, and **body** elements to get past total blank screens in both our application and my reproducer.  I've committed that change into the repo, so now you can see what it's looking like.  Here's what our app looks like with this change.
![MyApp - ScrollbarFix - 20230927-001](https://github.com/primefaces/community/assets/78611742/18373ddf-65e4-4a19-9d69-d6066cac9937)
It now has the scrollbar, but it no longer has the stuff across the top, a toolbar and menus, which are common to all our screens, and are defined in our template file, which is still referenced in what is now the the **html** tag.

How do I get back the stuff defined in our template file?

**Reply by melloware** [reply]
Without a reproducer I have no idea. Screenshots are something that doesn't really help us.

**Reply by NCIDevRonR** [reply]
Well I notice in our app that we declare our template within a ui:composition tag.

In a world where we're no longer using this tag, how should I declare our template going forward?

**Reply by melloware** [reply]
I think templates are OK and ui:composition is OK?  we use it in all our production apps.

**Reply by NCIDevRonR** [reply]
Ok.  I've added a ui:composition tag just inside the html tag.  Whenever I include a reference to the template file, I see the menus and top-of-screen stuff, but not the rest of the screen.  If I remove the template file reference, I can see the rest of the screen, but then I lose the menus et al.  This happens in our app, as well as in my [reproducer](https://github.com/NCIDevRonR/primefaces-test/tree/no-vertical-scrollbar).

How do I get both screen sections visible together?

**Reply by NCIDevRonR** [reply]
Since the vertical scrollbar thing is basically solved, but this template issue remains, I'm renaming this discussion after the latter issue.  I don't see anything documented in PF13-land about how to manage templates as part of a page.

**Reply by melloware** [reply]
Templates are a JSF concept not a PrimeFaces concept right?  If you mean `<ui:insert name="content" />` type templates or do you mean PrimeFaces premium Themes Templates?

**Reply by NCIDevRonR** [reply]
What we're carrying over from our PF6 implementation are templates declared like:  `<ui:composition template="filename.xhtml">`.

**Reply by melloware** [reply]
OK yeah that is JSF not really PF related as all the UI tags are core JSF either Mojarra or MyFaces.

**Reply by NCIDevRonR** [reply]
OK, then I'll chase down what I can find in the JSF docs about how it's all supposed to work.  I'll also look into how we might venture into PrimeFaces Premium and present that as a possible option to my team.  Thanks.

**Reply by NCIDevRonR** [reply]
Please refer again to [this reproducer](https://github.com/NCIDevRonR/primefaces-test/tree/no-vertical-scrollbar).  I have resolved my JSF template issue, and can once again see both sections of the screen.  With both sections visible, I have these two issues:

1. I can no longer see the scrollbar, even though I _think_ I still have the code changes from before which restored it.  Am I still missing something?
2. The menu is confined to the upper section of the screen, and is often cut off.  In our application, the menu needs to be able to flow down to the bottom of the screen, covering much of the main section.  Am I missing some PF switch to enable it to flow this way?

**Reply by melloware** [reply]
Sorry I don't have time to investigate. I have given you all the hints I can give for now.

**Reply by NCIDevRonR** [reply]
Ok, well your "hints" from before are still in place.  I'd be interested to know if there's a basic incompatibility in PF13 between the JSF template and the new layout panes which prevents the scrollbars from working.

If you're too busy to look at it, and if nobody else is going to chime in, I'll continue trying variations to see if I can get the scrollbars to work again and the menus to play nice with the rest of the screen.  I'll let you know if I'm able to figure it out.  These are the last two issues on my to-do list before I begin applying the fixes from here into our application.

**Reply by melloware** [reply]
Nothing that i know of and we heavily use PrimeFaces 13 and the PrimeFaces Extensions Showcase is on 13 all using Layout and Layout Panes so i know it works.  You will need to deep dive and figure out why yours is not working.

**Reply by NCIDevRonR** [reply, marked as answer]
Deep Dive Done!!  Refer to [this reproducer](https://github.com/NCIDevRonR/primefaces-test/tree/no-vertical-scrollbar).  Moving the pe:layoutPane tag from the "body" section of the main test page into the corresponding "body" section of the template file resolved my scrollbar issue.  This fix worked in our application as well, so soon I'll be replicating it to all 20 or so of our template-driven pages.

Since the menu cutting off issue occurs in my reproducer but not in our application, I'm happy to leave it as a question for the ages.

Thanks for your help with these issues.
